### PR TITLE
Fix windows

### PR DESCRIPTION
**Pull request by mattn**
fix on windows. `start` is not found on windows.
this change include some refactoring.

### CODE REVIEW

File: launcher/launcher_windows.go [hunk 1]
```diff
@@ -3,17 +3,10 @@
...
-func OpenURI(uri string) {
-	cmd := exec.Command("start", uri)
-	err := cmd.Start()
-	if err != nil {
-		fmt.Errorf("%s\n", err)
-	} else {
-		err = cmd.Wait()
-	}
+func OpenURI(uri string) error {
+	return exec.Command("rundll32", "url.dll,FileProtocolHandler", uri).Start()
```
Comment (zquestz) [inline launcher/launcher_windows.go]: Why not just use the "start" binary? This recognizes URI's and opens them in your default browser. Not sure why we need to call rundll32 here...

File: providers/providers.go [hunk 2]
```diff
@@ -27,18 +29,17 @@ func AddProvider(name string, provider Provider) {
...
 // Search builds a search URL and opens it in your browser.
-func Search(p string, q string, verbose bool) {
+func Search(p string, q string, verbose bool) error {
 	builder := Providers[p]
 
 	if builder != nil {
 		url := builder.BuildURI(q)
 		if verbose {
-			fmt.Printf("%s\n", url)
+			fmt.Fprintf(os.Stderr, "%s\n", url)
```
Comment (zquestz) [inline providers/providers.go]: I don't think that verbose output should go to Stderr, only Stdout.

File: providers/providers.go [hunk 3]
```diff
@@ -50,12 +51,5 @@ func DisplayProviders() string {
...
 	sort.Strings(names)
-
-	output := ""
-
-	for _, p := range names {
-		output = output + fmt.Sprintf("%s\n", p)
-	}
-
-	return output
+	return strings.Join(names, "\n") + "\n"
```
Comment (zquestz) [inline providers/providers.go]: Nice catch, I had other formatting here I removed and this is much better.

File: providers/providers.go [hunk 4]
```diff
@@ -27,18 +29,17 @@ func AddProvider(name string, provider Provider) {
...
 	if builder != nil {
 		url := builder.BuildURI(q)
 		if verbose {
-			fmt.Printf("%s\n", url)
+			fmt.Fprintf(os.Stderr, "%s\n", url)
 		}
-		launcher.OpenURI(url)
-	} else {
-		fmt.Printf("Provider %q not supported!\n", p)
+		return launcher.OpenURI(url)
```
Comment (zquestz) [inline providers/providers.go]: While this reduces some boilerplate, we no longer wait for the command to finish. Not important for the "open" case, but eventually if we fork a process like w3m it won't work right. I want a way to be able to pass a binary to block on, so we can support terminal browsers.

**Comment by zquestz** [comment]
So according to Microsoft all current versions of supported windows should have "start"

https://technet.microsoft.com/en-us/library/cc770297.aspx

**Comment by mattn** [comment]
`start` is internal command on `cmd.exe` not executable.

**Comment by mattn** [comment]
If you want to use `start`, it should be `cmd /k start URL`.

**Comment by zquestz** [comment]
ahh gotcha, we should probably just use that. =)

**Comment by zquestz** [comment]
Lets fixup the PR just to refactor the start command for windows for now. The other refactor around OpenURI can wait till I have a way to customize the binary you call.

**Comment by mattn** [comment]
okay, will fix above except `start`.

File: providers/providers.go (hunk 4 above)
Comment (mattn) [inline providers/providers.go]: Then you MUST NOT use `cmd.Start`. Should use `cmd.Run` instead.

**Comment by mattn** [comment]
done. please take a look.

**Comment by mattn** [comment]
I'm not make sure behavior of xdg-open.

**Comment by mattn** [comment]
> ahh gotcha, we should probably just use that. =)

`cmd /k start URL` has problems which can't pass % encoded correctly. so it's better to use rundll32 instead.

**Comment by zquestz** [comment]
From what I read you have to quote the command. Does it work for you with:

```
cmd /k "start <url>"
```

?

**Comment by mattn** [comment]
When user pass argument contains double quote, it will break.

**Comment by mattn** [comment]
And `"%FOO%"` will be decoded like `"$FOO"` on UNIX OSs.

**Comment by zquestz** [comment]
hmmm the docs state that double quotes around the entire command should work...

**Comment by zquestz** [comment]
try:

```
cmd /k "start https://google.com"
```

**Comment by zquestz** [comment]
Btw I have a refactor that allows the user to specify a custom binary. Mainly to support terminal browsers like w3m. Will be pushing that shortly and it will affect your PR.

**Comment by mattn** [comment]
As I said, `cmd /k "start ..."` can't pass argument correctly.



> Btw I have a refactor that allows the user to specify a custom binary. Mainly to support terminal browsers like w3m. Will be pushing that shortly and it will affect your PR.

Okay, but my change is already possible to customize browser with set S_BROWSER environment variable. And it works with w3m.

**Comment by zquestz** [comment]
Ahh I see now, thanks for clarifying. Bringing up a Windows box now so I can do some testing.

**Comment by zquestz** [comment]
K just pushed my build. Need to adjust to fix windows. VM should be up soon.

**Comment by zquestz** [comment]
So in your screenshot it did parse the args correctly. Passed 1ab which is correct. Give this build a try and let me know.

**Comment by mattn** [comment]
currently, my patch is heavy conflicted. I'll make this PR again.

**Comment by mattn** [comment]
done. please take a look.

**Comment by zquestz** [comment]
Yeah I cleaned up a bunch of stuff. Exit codes are all now working correctly too. I think I just need to pull in your dll command at this point.

**Comment by mattn** [comment]
Sorry, I had tired to rebase master branch several times. please do it your self.

**Comment by zquestz** [comment]
Thanks again for helping me get the right command for windows. Your fix has been incorporated into master. Sorry for the frustration, codebase is still moving around a bit.

**Pull request closed without merging**